### PR TITLE
[3.2] Optimized physics object spawn time

### PR DESCRIPTION
**Pull request by aaronfranke**
Backported to 3.2 from pull request #39726 from AndreaCatania/add_body_impr_physics

Tested on the Voxel demo project from the Godot demo projects repo.

### CODE REVIEW

**Comment by akien-mga** [comment]
FYI, when cherry-picking commits I prefer using `git cherry-pick -x <commit hash>` and then resolve the conflicts, which will preserve the commit log and author information, and add a reference to the original commit hash in the commit description.

**Comment by aaronfranke** [comment]
I'll keep that in mind for next time. If the changes conflict too much, a lot of the time I just try to get through the conflict resolution tool as soon as possible, then I fix the code manually and squash. Perhaps it's not the best approach :P

**Comment by akien-mga** [comment]
Yeah I don't use any conflict resolution tool or strategy myself, I just open all conflicting files in `nano` or `kate` and fix conflicts manually.

**Comment by madmiraal** [comment]
Note: #40311 is a regression in #39726 that needs to be resolved before merging this.

**Comment by aaronfranke** [comment]
#40252 solved #40311, but it caused #40840 and #40848, and the pending fix for the former is https://github.com/godotengine/godot/pull/40886.

File: modules/bullet/collision_object_bullet.cpp [hunk 1]
```diff
@@ -329,59 +354,71 @@ bool RigidCollisionObjectBullet::is_shape_disabled(int p_index) {
...
-	bulletdelete(shp.bt_shape);
+	shp.release_bt_shape();
 	reload_shapes();
 }
 
 void RigidCollisionObjectBullet::reload_shapes() {
+	need_shape_reload = true;
+	if (space) {
+		space->add_to_flush_queue(this);
+	}
```
Comment (AndreaCatania) [inline modules/bullet/collision_object_bullet.cpp:378]: The shape should not be reloaded during the flush but during the pre_process. To fix the issue #40840 you need to insert `p_body->pre_process();` at the beginning of the function `test_body_motion`.

File: modules/bullet/rigid_body_bullet.cpp [hunk 2]
```diff
@@ -343,23 +341,24 @@ void RigidBodyBullet::set_space(SpaceBullet *p_space) {
...
 	if (space) {
-		space->add_rigid_body(this);
+		space->register_collision_object(this);
+		reload_body();
+		space->add_to_flush_queue(this);
 	}
 }
 
 void RigidBodyBullet::dispatch_callbacks() {
+	RigidCollisionObjectBullet::dispatch_callbacks();
```
Comment (AndreaCatania) [inline modules/bullet/rigid_body_bullet.cpp:358]: This is not needed. Check this: https://github.com/godotengine/godot/pull/40886#issuecomment-666964762

**Review by AndreaCatania — CHANGES_REQUESTED** [review]
Thanks for this PR, just few things to change.

**Comment by AndreaCatania** [comment]
Would be nice port `LocalVector`, and use it here too. So to have feature parity between 3.2 and the 4.0. This should also remove the need to cast the value returned by `size`: `for (uint32_t i = 0; i < uint32_t(areasWhereIam.size()); i += 1) {`

**Comment by aaronfranke** [comment]
If we need to add core engine features to backport this feature, then it might be better to just not backport this, and simply require that people wait for 4.0 if they want improved physics object spawn times.

**Comment by AndreaCatania** [comment]
Port `LocalVector` is just matter of insert it into the core directory. Then, you can use it into the bullet. It's not super needed, but if done you can remove a lot of overhead.

Feel free to do it, or not.

File: modules/bullet/collision_object_bullet.cpp (hunk 1 above)
Comment (aaronfranke) [reply modules/bullet/collision_object_bullet.cpp:378]: I don't understand what you mean. I already added `p_body->pre_process();` in `test_body_motion` on the line you mentioned in this comment https://github.com/godotengine/godot/pull/40886#issuecomment-666964762 (see https://github.com/godotengine/godot/pull/40188/commits/24d50fde7bb5bc528cb916a4b7d48fced7adff10#diff-a61d01bcaef4874b6713c0b00e6d4571R1014)

File: modules/bullet/rigid_body_bullet.cpp (hunk 2 above)
Comment (aaronfranke) [reply modules/bullet/rigid_body_bullet.cpp:358]: So you want me to delete the call to `RigidCollisionObjectBullet::dispatch_callbacks()` inside of `RigidBodyBullet::dispatch_callbacks()`? I don't understand what that comment has to do with this line.

**Comment by aaronfranke** [comment]
Because of the numerous regressions caused by #39726, #40252, and maybe others, I'm closing this. It would have been nice to have in 3.2, since it significantly improves performance in the Voxel demo, but there are just too many regressions.

I'm just going to make a blanket statement that the dumpster fire known as Godot's physics is far too risky to take cherry-picks from, and from now on I would suggest all physics changes be kept as 4.0+ only.

**Comment by AndreaCatania** [comment]
After all the work done here, I would not close this. But well, is up to you I think.

**Pull request closed without merging**